Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift

```diff
@@ -540,6 +563,45 @@ private extension CardPresentPaymentStore {
         storage.deleteObject(storageAccount)
         storage.saveIfNeeded()
     }
+
+    func upsertChargeInBackground(readonlyCharge: WCPayCharge) {
+        let storage = storageManager.viewStorage
+        let storageWCPayCharge = storage.loadWCPayCharge(siteID: readonlyCharge.siteID, chargeID: readonlyCharge.id) ??
+        storage.insertNewObject(ofType: Storage.WCPayCharge.self)
+
+        switch readonlyCharge.paymentMethodDetails {
+        case .cardPresent(let details):
+            let storageCardPresentDetails = storageWCPayCharge.cardPresentDetails ??
+            storage.insertNewObject(ofType: Storage.WCPayCardPresentPaymentDetails.self)
+            let storageReceiptDetails = storageCardPresentDetails.receipt ?? storage.insertNewObject(ofType: Storage.WCPayCardPresentReceiptDetails.self)
+            storageCardPresentDetails.update(with: details)
+            storageReceiptDetails.update(with: details.receipt)
+            storageCardPresentDetails.receipt = storageReceiptDetails
+            storageWCPayCharge.cardPresentDetails = storageCardPresentDetails
+            storageWCPayCharge.cardDetails = nil
+        case .card(let details):
+            let storageCardDetails = storageWCPayCharge.cardDetails ?? storage.insertNewObject(ofType: Storage.WCPayCardPaymentDetails.self)
+            storageCardDetails.update(with: details)
+            storageWCPayCharge.cardDetails = storageCardDetails
+            storageWCPayCharge.cardPresentDetails = nil
+        case .unknown:
+            storageWCPayCharge.cardDetails = nil
+            storageWCPayCharge.cardPresentDetails = nil
+            break
+        }
```

Comment:
I wonder if all of this should be done inside the `WCPayCharge.update` method instead. I don't have a strong opinion, but it seems like it'd be my first choice.

Reply:
None of the existing `ReadOnlyConvertible` `update(with:)` functions that I can see do anything more complicated than translating enums to their `rawValue`... where there are arrays of other entities on the `Networking.model`, the `toReadOnly` function (part of the same protocol) will map those arrays and call `toReadOnly` on them, but it won't convert them the other way in the `update` method.

There's no calls to `insertNewObject` in any of those methods, and I think it'd feel pretty against the grain to do that there in this one.

It is a bit of a wall of text with some long lines, so I'm going to break it up a little.